src/metabase/task/send_pulses.clj

```diff
@@ -86,20 +98,33 @@
     (t2/delete! :model/PulseChannel :id [:in ids-to-delete])
     (set ids-to-delete)))
 
-(defn- clear-pcs-and-send-pulse!
-  [pulse-id channel-ids]
+(defn- send-pulse!*
+  "Do several things:
+  - Clear PulseChannels that have no recipients and no channel set for a pulse
+  - Send a pulse to a list of channels
+  - Update the priority of the trigger if the pulse is sent successfully"
+  [schedule-map pulse-id channel-ids]
   (let [cleared-channel-ids         (clear-pulse-channels-no-recipients! pulse-id)
         to-send-channel-ids         (set/difference channel-ids cleared-channel-ids)
         to-send-enabled-channel-ids (t2/select-pks-set :model/PulseChannel :id [:in to-send-channel-ids] :enabled true)]
     (if (seq to-send-enabled-channel-ids)
-     (send-pulse! pulse-id to-send-enabled-channel-ids)
-     (log/infof "Skip sending pulse %d because all channels have no recipients" pulse-id))))
+      (let [start    (System/currentTimeMillis)
+            result   (send-pulse! pulse-id to-send-enabled-channel-ids)
+            end      (System/currentTimeMillis)
+            ;; we set priority as duration in seconds, the quicker the pulse is sent the higher the priority
+            ;; highest priorities triggers are executed first, so we need to invert it with the duration
+            priority (-> (- end start) (/ 1000) int -)]
```

Comment:
If priorities are negative, that means pulses will be prioritised after all the other tasks like sync. Is that a conscious decision?

Reply:
right, that's not ideal. the default priority is 5 so we should have a priority higher than 5. what's the best way to ensure that 🤔

Reply:
Many ways, but a suggestion to keep it simple:

if `(- end start)` is less than 1 minute: default + 3
if `(- end start)` is less than 10 minute: default + 2
otherwise: default + 1

Reply:
how about this: https://github.com/metabase/metabase/pull/42331/commits/be9fe890e9fa545397e80bfe1f35fb6eefb69aa7?

Reply:
yes, it's good :D

Reply:
if a task takes more than an hour to send pulse then it deserves to get a lower priority than sync I guess haha

Reply:
`(->> (int (/ duration 1000)) (min 1000) (/ 1) (* 995) (+ 5))`

It's kinda weird in clojure, but it's `(1 / min(duration-secs, 1000)) * scale (995 if it's from 5 to 1000) + min-priority (5)`.

Reply:
ah, giving a `min` is a good idea.
I still think giving a maximum duration is easier to reason tho :p.

```clojure
(-> ms-duration
      (/ 1000)
      ;; assuming 1 hour is the max duration a pulse can take
      (#(- (* 60 60) %))
      ;; make sure it's higher than 5 (the default priority)
      (min 6))
```

Reply:
Yeah; it just math sucks a bit in lisp syntax; I wish Clojure had a little `math` macro or something, honestly :)